pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeSegmentDataManager.java

```diff
@@ -703,9 +703,11 @@ public void run() {
         //   persisted.
         // Take upsert snapshot before starting consuming events
         if (_partitionUpsertMetadataManager != null) {
-          _partitionUpsertMetadataManager.takeSnapshot();
-          // If upsertTTL is enabled, we will remove expired primary keys from upsertMetadata after taking snapshot.
+          // we should remove keys first and then take snapshot, this is because deletedKeysTTL flow removes keys from
```

Comment:
this will be incorrect behavior for upsert ttl. Can you please check whether deleteKeyTTL is enabled before changing the order.

Reply:
Hey @deemoliu i have updated based on your suggestion! Can you also elaborate on why is this necessary? I can maybe update javadocs as well. Going through `metadataTTL` code i didn't find any strict requirement for taking upsertSnapshot first but i might be definitely missing something.

Reply:
Updated javadocs as well!